internal/destination/destination.go

```diff
@@ -131,29 +135,44 @@ func (d *Destination) Open(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("connect to NATS: %w", err)
 	}
-
-	d.writer, err = NewWriter(WriterParams{
-		Conn:          conn,
-		Subject:       d.config.Subject,
-		RetryWait:     d.config.RetryWait,
-		RetryAttempts: d.config.RetryAttempts,
-	})
-	if err != nil {
-		return fmt.Errorf("init jetstream writer: %w", err)
-	}
+	d.nc = conn
 
 	// Async handlers & callbacks
 	conn.SetErrorHandler(internal.ErrorHandlerCallback(ctx))
-	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {}))
-	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {}))
+	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {
+		d.writer.stopWrites()
+	}))
+	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {
+		d.writer, err = NewWriter(writerParams{
+			nc:            d.nc,
+			subject:       d.config.Subject,
+			retryWait:     d.config.RetryWait,
+			retryAttempts: d.config.RetryAttempts,
+		})
+		d.writer.startWrites()
+	}))
 	conn.SetClosedHandler(internal.ClosedCallback(ctx))
 	conn.SetDiscoveredServersHandler(internal.DiscoveredServersCallback(ctx))
 
+	d.writer, err = NewWriter(writerParams{
+		nc:            d.nc,
+		subject:       d.config.Subject,
+		retryWait:     d.config.RetryWait,
+		retryAttempts: d.config.RetryAttempts,
+	})
+	if err != nil {
+		return fmt.Errorf("init jetstream writer: %w", err)
+	}
+
 	return nil
 }
 
 // Write writes a record into a Destination.
 func (d *Destination) Write(_ context.Context, records []sdk.Record) (int, error) {
+	if !d.nc.IsConnected() {
+		return 0, errWriteUnavailable
+	}
```

Comment:
This problem still exists, I'm pretty sure this will cause the pipeline to stop with an error if Conduit tries to write a record while the connector is reconnecting (not the behavior we want, right?). We should block here until the connection is established _or_ until the context is canceled, whatever happens first. See for example how we are doing this in the [grpc-client connector](https://github.com/conduitio-labs/conduit-connector-grpc-client/pull/9/files#diff-b0e314d1e24c22a62ecb712e94913d77bed15abfd75e80a0b61189c0abe17453R198-R222).

Reply:
I added a retry/wait logic around the function.